gslib/utils/copy_helper.py

```diff
@@ -670,59 +673,67 @@ def ConstructDstUrl(src_url,
   #    should create the objects gs://bucket/f1.txt and gs://bucket/f2.txt,
   #    assuming dir1 contains f1.txt and f2.txt.
 
-  recursive_move_to_new_subdir = False
+  # If there are multiple top-level source items, copy full file hierarchies.
+  # This avoids conflicts such as "dir1/f.txt" and "dir2/f.txt" both getting
+  # copied to "gs://bucket/f.txt". Linux normally errors on these conflicts,
+  # but we cannot do that because we need to give users the ability to create
+  # dirs as they copy to the cloud.
+  # Ignore this rule if using double wildcard ** because that treats all files
+  # as top-level, in which case the user doesn't want to preserve directories.
+  preserve_src_top_level_dirs = '**' not in src_url.versionless_url_string and (
+      has_multiple_top_level_srcs or have_existing_dest_subdir)
   if (global_copy_helper_opts.perform_mv and recursion_requested and
```

Comment:
Is this still specific to mv? I thought we were planning to change this behavior for cp as well?

Reply:
Right, but what I am saying is that wouldn't the behavior for mv and cp end up being the same with the new naming rules? do we have to treat mv and cp differently?

Reply:
I don't think the behavior is the same. This block is more like "rename" behavior.
More details in responses below.

Reply:
Discussed offline. Waiting for Nick to provide command examples/scenarios where mv and cp will have different behavior.

Reply:
Removing Case I looks like a success